### PR TITLE
Implement reading of configuration properties files in Core SDK

### LINKED ISSUE #597: Enhance Core SDK to support .properties configuration
The Core SDK should allow users to specify `.properties` in the following locations to override defaults.

1. `classpath:com/stormpath/sdk/config/stormpath.properties`
2. `classpath:stormpath.properties`
3. `$HOME/.stormpath/apiKey.properties`
4. `$HOME/.stormpath/stormpath.properties`
5. `System.getProperty("user.dir") + File.separator + "apiKey.properties"`
6. `System.getProperty("user.dir") + File.separator + "stormpath.properties`

NOTE: The first file does not currently exist, so it will need to be created with all the [defaults](https://github.com/stormpath/stormpath-sdk-spec/blob/master/specifications/config.md#default-configuration).

See https://github.com/stormpath/stormpath-sdk-java/wiki/Overriding-Default-Configuration for additional information.

### PR DESCRIPTION
**Pull request by mraible**
Resolves #597

Implement reading of configuration properties files in the following order:

1. classpath:com/stormpath/sdk/config/stormpath.properties
2. classpath:stormpath.properties
3. ~/.stormpath/stormpath.properties
4. ~/stormpath.properties

This is a work-in-progress; I wanted to show my approach to see if it's the best way to do this.

### CODE REVIEW

File: impl/src/main/java/com/stormpath/sdk/impl/client/DefaultClientBuilder.java [hunk 1]
```diff
@@ -39,15 +50,104 @@
...
+    public DefaultClientBuilder() {
+        for (String location : DEFAULT_STORMPATH_PROPERTIES_FILE_LOCATIONS) {
+            Properties props = getPropertiesFromFile(location);
+
+            // check to see if property value is null before setting value
+            // if != null, allow it to override previously set values
+            if (getPropertyValue(props, DEFAULT_CLIENT_API_KEY_FILE_PROPERTY_NAME) != null) {
+                apiKeyFile = getPropertyValue(props, DEFAULT_CLIENT_API_KEY_FILE_PROPERTY_NAME);
+            } else {
+                // todo: support json and yaml overrides
```
Comment (mrioan) [inline impl/src/main/java/com/stormpath/sdk/impl/client/DefaultClientBuilder.java]: I guess there is an issue already created for this TODO. Can you please reference it here?

File: impl/src/main/java/com/stormpath/sdk/impl/client/DefaultClientBuilder.java [hunk 2]
```diff
@@ -39,15 +50,104 @@
...
+    private static final String USER_HOME = System.getProperty("user.home") + File.separatorChar;
+    private static final String STORMPATH_PROPERTIES = "stormpath.properties";
+    private static final String[] DEFAULT_STORMPATH_PROPERTIES_FILE_LOCATIONS = {
+            ClasspathResource.SCHEME_PREFIX + "com/stormpath/sdk/config/" + STORMPATH_PROPERTIES,
+            ClasspathResource.SCHEME_PREFIX + STORMPATH_PROPERTIES,
+            USER_HOME + ".stormpath" + File.separatorChar + STORMPATH_PROPERTIES,
+            USER_HOME + STORMPATH_PROPERTIES
+    };
+
+    private String apiKeyFile;
```
Comment (mrioan) [inline impl/src/main/java/com/stormpath/sdk/impl/client/DefaultClientBuilder.java]: IntelliJ says that `apiKeyFile` is never used

File: impl/src/main/java/com/stormpath/sdk/impl/client/DefaultClientBuilder.java [hunk 3]
```diff
@@ -39,15 +50,104 @@
...
+    private String cacheManagerCaches;
+    private String baseUrl;
+    private int connectionTimeout;
     private AuthenticationScheme authenticationScheme;
-    private CacheManager         cacheManager;
+    private int proxyPort;
+    private String proxyHost;
+    private String proxyUsername;
+    private String proxyPassword;
+    private String applicationName;
```
Comment (mrioan) [inline impl/src/main/java/com/stormpath/sdk/impl/client/DefaultClientBuilder.java]: IntelliJ says that `applicationName` and `applicationHref` are never used

**Comment by mrioan** [comment]
@mraible Code looks good overall. However I would like to see a few tests validating that values are been overwritten in the proper order

**Comment by mraible** [comment]
@mrioan Thanks for taking a look. I'll definitely add more tests. 

One question: where should the default applicationName and applicationHref be used?

**Comment by mrioan** [comment]
@mraible the application name and href can be removed from Core SDK properties. They were erroneously kept in the spec but are actually not needed. Thanks!

**Comment by mraible** [comment]
@mrioan I have removed the application name and href as properties. Please review again. 
/cc @lhazlewood @dogeared

**Comment by mraible** [comment]
I noticed this morning that the spec has validation notes about `application.name` and `application.href`. Should that section be removed from the spec?

https://github.com/stormpath/stormpath-sdk-spec/blob/master/specifications/config.md#application

**Comment by lhazlewood** [comment]
You're right @mraible - I'll mention that to the SDK teams today.

**Comment by mraible** [comment]
FWIW, code coverage for `DefaultClientBuilder` is at 91%:

https://afternoon-oasis-83667.herokuapp.com/com/stormpath/sdk/impl/client/DefaultClientBuilder.html

**Comment by lhazlewood** [comment]
Nice!

**Comment by mraible** [comment]
As part of #588, I added a test to confirm that the default spec config matches the default properties config. 

https://github.com/stormpath/stormpath-sdk-java/pull/607/files#diff-dc0a42f21b7640d261c988a10996018bR1

**Pull request closed without merging**